lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,539 @@
+  if (!parsedLocaleIdentifier.transformedLanguageIdentifier.languageSubtag.empty() ||
+      !parsedLocaleIdentifier.transformedExtensionFields.empty()) {
+    return false;
+  }
+  
+  bool hasExtension = false;
+  if (isUnicodeLanguageSubtag(getCurrentSubtagRef())) {
+    hasExtension = true;
+    parseUnicodeLanguageId(true);
+    if (!hasMoreSubtags()) {
+      return true;
+    }
+  } 
+
+  if (isTransformedExtensionKey(getCurrentSubtagRef())) {
+    hasExtension = true;
+
+    while (true) {
+      if (!isTransformedExtensionKey(getCurrentSubtagRef())) {
+        break;
+      }
+      
+      std::u16string tkey = getCurrentSubtag();
+      
+      std::vector<std::u16string> &tvalues = parsedLocaleIdentifier.transformedExtensionFields.insert({tkey, {}}).first->second;
+      
+      // read key then one or more values
+      if (!nextSubtag()) {
+        return false;
+      }
+      if (!isTransformedExtensionTValueItem(getCurrentSubtagRef())) {
+        return false;
+      }
+      tvalues.push_back(getCurrentSubtag());
+      if (!nextSubtag()) {
+        return true;
+      }
+      while (true) {
+        if (!isTransformedExtensionTValueItem(getCurrentSubtagRef())) {
+          break;
+        }
+        tvalues.push_back(getCurrentSubtag());
+        if (!nextSubtag()) {
+          return true;
+        }
+      }
+    }
+  } 
+
+  return hasExtension;
+}
+
+// pu_extensions= sep [xX]
+// (sep alphanum{1,8})+ ;
+//
+// No tokens may appear after pu_extensions
+bool LanguageTagParser::parsePUExtension() {
+  if (!isPrivateUseExtension(getCurrentSubtagRef())) {
+    return false;
+  }
+  parsedLocaleIdentifier.puExtensions.push_back(getCurrentSubtag());
+
+  if (!nextSubtag()) {
+    return true;
+  }
+
+  while (isPrivateUseExtension(getCurrentSubtagRef())) {
+    parsedLocaleIdentifier.puExtensions.push_back(getCurrentSubtag());
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+
+  // Tokens are not expected after pu extension
+  return false;
+}
+
+// other_extensions= sep [alphanum-[tTuUxX]]
+// (sep alphanum{2,8})+ ;
+bool LanguageTagParser::parseOtherExtension(char16_t singleton) {
+  std::unordered_map<char16_t, std::vector<std::u16string>> &extMap = parsedLocaleIdentifier.otherExtensionMap;
+  if (extMap.find(singleton) != extMap.end()) {
+    return false;
+  }
+  
+  extMap.insert({singleton, {}});
+  std::vector<std::u16string> *otherExtensions = &extMap.find(singleton)->second;
+
+  if (!isOtherExtension(getCurrentSubtagRef())) {
+    return false;
+  }
+  otherExtensions->push_back(getCurrentSubtag());
+  if (!nextSubtag()) {
+    return true;
+  }
+
+  while (isOtherExtension(getCurrentSubtagRef())) {
+    otherExtensions->push_back(getCurrentSubtag());
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+  
+  return true;
+}
+
+bool LanguageTagParser::hasMoreSubtags() {
+  return localeId_.length() > 0 && subtagEnd_ < (int)localeId_.length() - 1;
+}
+bool LanguageTagParser::nextSubtag() {
```

Comment:
I think much of this code would become simpler if `subtagEnd_` was instead one past the end of a subtag.